### ISSUE TITLE
Run DDP without torch.distributed.launch --nproc_per_node ?

### ISSUE THREAD

**Issue by pngmafia** [ultralytics/yolov5#2583]
I am writing a custom training script in which I cannot give `torch.distributed.launch --nproc_per_node` options in a python command. Is it possible to run DDP without `torch.distributed.launch --nproc_per_node` options, if so what are the changes to be done to train.py for it to run on DDP.
please help

thanks

**Comment by glenn-jocher** [comment]
@pngmafia if you're willing to commit all CUDA_VISIBLE_DEVICES to your training run then you could use some bash trickery to count the number of devices and return that value. Someone smarter than me would have to fill in the details here but the idea is pretty simple:

```
$ python -m torch.distributed.launch --nproc_per_node $(count number of $CUDA_VISIBLE_DEVICES) train.py
```

**Comment by glenn-jocher** [comment]
See https://askubuntu.com/questions/934963/how-to-get-number-of-gpu-cards-i-have-from-a-command-line
```
nvidia-smi --query-gpu=name --format=csv,noheader | wc -l
```

**Comment by pngmafia** [comment]
No, I am writing a custom script that triggers the train function. So it's not possible to run python command at all. I can just run python train.py(without `torch.distributed.launch --nproc_per_node`). 
Just running python train.py should trigger DDP, for that to happen what all are the changes I have to make in train.py. 
Does my question make sense?

**Comment by glenn-jocher** [comment]
@pngmafia no your question does not make sense, the commands I've provided are not python commands, they are bash commands, and DDP is not applied by calling train.py directly. You may want to start by having a look at the Multi-GPU training tutorial:

## YOLOv5 Tutorials

* [Train Custom Data](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data)&nbsp; 🚀 RECOMMENDED
* [Weights & Biases Logging](https://github.com/ultralytics/yolov5/issues/1289)&nbsp; 🌟 NEW
* [Multi-GPU Training](https://docs.ultralytics.com/yolov5/tutorials/multi_gpu_training)
* [PyTorch Hub](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading)&nbsp; ⭐ NEW
* [ONNX and TorchScript Export](https://docs.ultralytics.com/yolov5/tutorials/model_export)
* [Test-Time Augmentation (TTA)](https://docs.ultralytics.com/yolov5/tutorials/test_time_augmentation)
* [Model Ensembling](https://docs.ultralytics.com/yolov5/tutorials/model_ensembling)
* [Model Pruning/Sparsity](https://docs.ultralytics.com/yolov5/tutorials/model_pruning_and_sparsity)
* [Hyperparameter Evolution](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution)
* [Transfer Learning with Frozen Layers](https://docs.ultralytics.com/yolov5/tutorials/transfer_learning_with_frozen_layers)&nbsp; ⭐ NEW
* [TensorRT Deployment](https://github.com/wang-xinyu/tensorrtx)

**Comment by pngmafia** [comment]
Does hard coding these values help? if so what should. they be hard coded to?
https://github.com/ultralytics/yolov5/blob/e5b0200cd250759c782207160761ca9756300065/train.py#L492
https://github.com/ultralytics/yolov5/blob/e5b0200cd250759c782207160761ca9756300065/train.py#L476
https://github.com/ultralytics/yolov5/blob/e5b0200cd250759c782207160761ca9756300065/train.py#L491
Iam always going to use all the gpus available and will be providing the number of gpus as input.

**Comment by glenn-jocher** [comment]
@pngmafia local rank is never hard codeable as it will vary by process. I would recommend you follow the Multi-GPU Training tutorial for best practices when training.

## YOLOv5 Tutorials

* [Multi-GPU Training](https://docs.ultralytics.com/yolov5/tutorials/multi_gpu_training)

**Comment by pngmafia** [comment]
@glenn-jocher 
ok, let me reframe my question. I don't want to give `torch.distributed.launch` or `--nproc_per_node` in my command. I just want to give python train.py .....(followed by the training options). What are the changes I can make in train.py for it to still run DDP. I see in train.py you take the 'RANK' and 'WORLD_SIZE' from an environment variable does hardcoding them allow me to run the training on DDP?

**Comment by pngmafia** [comment]
@glenn-jocher I see local_rank being set to -1 here https://github.com/ultralytics/yolov5/blob/e5b0200cd250759c782207160761ca9756300065/train.py#L476
and it is not being updated anywhere else. How does it vary by the process?

**Comment by glenn-jocher** [comment]
I would recommend you follow the Multi-GPU Training tutorial for best practices when training.

## YOLOv5 Tutorials

* [Multi-GPU Training](https://docs.ultralytics.com/yolov5/tutorials/multi_gpu_training)

**Comment by NanoCode012** [comment]
Reading this made me recall a possible solution. If someone could rewrite the code base slightly to use multiprocessing.spawn , we can do away with calling torch.distributed.launch . I don’t have the time now to do it, but I’m leaving it here for future readers.

**Comment by glenn-jocher** [comment]
@NanoCode012 yeah I remember your suggestion from before. Is this really possible though? Do you know of any torch documentation on this other method? 

I don't have time to do this now but it would be nice to unify the training command, i.e. basically so the current DP command would actually launch DDP training. Maybe the best we could do for now is create an issue for the idea with any documentation available and I'll put a TODO label on it. The TODO issues tend to get taken care of eventually as I review them all periodically.

We started using AWS spot instances recently that restart often, so we wrote an init script for these machines which calls resume.py, a new helper file I wrote that will restart all interrupted trainings present in a yolov5/ directory, for any combination of single-GPU, multi-GPU, and multi-GPU with different GPU counts.
https://github.com/ultralytics/yolov5/blob/master/utils/aws/resume.py

**Comment by NanoCode012** [comment]
@glenn-jocher , you can find it easily if you lookup `torch multiprocessing spawn`. I added a link below for reference.
https://pytorch.org/docs/stable/notes/multiprocessing.html

There were quite a few problems when I attempted before such as lower map(seed error) and worker always being recreated(Infinite Dataloader should solve this) and slower speeds. Maybe if someone can try it again, it should be smoother.

**Comment by glenn-jocher** [comment]
@NanoCode012 yes you make good points. The road to successful DDP updates is fraught with dangers.

Actually a much more useful Multi-GPU update would be to test.py. test.py is still stuck in single-GPU mode at all times, even if train.py runs DDP. For some types of runs this can become a bit of a bottleneck (i.e. the argoverse HD test set is very large compared to the train set, unlike COCO), i.e. 8-GPU DDP when training, but 1-GPU with 7 idle when testing.

Originally this was due to FP16 inference problems with DP I had when casting the model to .half(), but I think now we can use autocast with mixed precision inference to at least use DP with test.py.

**Comment by PyNancy** [comment]
@glenn-jocher  hi , i want to known current code support multi-machine with multi-gpu training ? can you offer me training command or args for multi-machine with multi-gpu training if current code has supported? 
thanks a lot!

**Comment by glenn-jocher** [comment]
@PyNancy yes, see Multi-GPU Training tutorial:

## YOLOv5 Tutorials

* [Train Custom Data](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data)&nbsp; 🚀 RECOMMENDED
* [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/)&nbsp; ☘️ RECOMMENDED
* [Weights & Biases Logging](https://github.com/ultralytics/yolov5/issues/1289)&nbsp; 🌟 NEW
* [Supervisely Ecosystem](https://docs.ultralytics.com/yolov5/tutorials/model_export8)&nbsp; 🌟 NEW
* [Multi-GPU Training](https://docs.ultralytics.com/yolov5/tutorials/multi_gpu_training)
* [PyTorch Hub](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading)&nbsp; ⭐ NEW
* [TorchScript, ONNX, CoreML Export](https://docs.ultralytics.com/yolov5/tutorials/model_export) 🚀
* [Test-Time Augmentation (TTA)](https://docs.ultralytics.com/yolov5/tutorials/test_time_augmentation)
* [Model Ensembling](https://docs.ultralytics.com/yolov5/tutorials/model_ensembling)
* [Model Pruning/Sparsity](https://docs.ultralytics.com/yolov5/tutorials/model_pruning_and_sparsity)
* [Hyperparameter Evolution](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution)
* [Transfer Learning with Frozen Layers](https://docs.ultralytics.com/yolov5/tutorials/transfer_learning_with_frozen_layers)&nbsp; ⭐ NEW
* [TensorRT Deployment](https://github.com/wang-xinyu/tensorrtx)

**Comment by PyNancy** [comment]
@NanoCode012  ok，i see it, thanks a lot

**Comment by ndvbd** [comment]
How to run the 
`python -m torch.distributed.launch --nproc_per_node=1 [SCRIPTNAME]`
inside an IDE with debugging?

**Comment by glenn-jocher** [comment]
@ndvbd in order to run the `python -m torch.distributed.launch --nproc_per_node=1 [SCRIPTNAME]` command inside an IDE with debugging, you can follow these steps:

1. Open your IDE (e.g., PyCharm, Visual Studio Code).

2. Create a new Python run configuration for your script.

3. In the run configuration settings, modify the "Script path" to include the command `python -m torch.distributed.launch --nproc_per_node=1 [SCRIPTNAME]`. 

4. Save the run configuration and start the debugging session.

By modifying the script path in the run configuration settings, you can run the command directly in the IDE while also having access to the debugging features provided by the IDE.

Please let me know if you need further assistance or have any other questions.

**Comment by ndvbd** [comment]
Thank you! It works, but I need to change it to run "module" instead of "script", and then specify "torch.distributed.launch" as module name and all the rest in the parameters.

**Comment by glenn-jocher** [comment]
@ndvbd great to hear that it works for you! To run "module" instead of "script" using `torch.distributed.launch`, you can modify the command as follows:

```
python -m torch.distributed.launch --nproc_per_node=1 -m [MODULENAME] [PARAMETERS]
```

Simply replace `[MODULENAME]` with the module name you want to run and `[PARAMETERS]` with any additional command-line parameters required by the module.

Feel free to reach out if you have any further questions or need additional assistance.
